src/main/java/org/opensearch/knn/index/memory/breaker/NativeMemoryCircuitBreakerService.java

```diff
@@ -0,0 +1,201 @@
+import org.opensearch.common.unit.ByteSizeValue;
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.index.memory.NativeMemoryCacheManager;
+import org.opensearch.knn.plugin.stats.StatNames;
+import org.opensearch.knn.plugin.transport.KNNStatsAction;
+import org.opensearch.knn.plugin.transport.KNNStatsNodeResponse;
+import org.opensearch.knn.plugin.transport.KNNStatsRequest;
+import org.opensearch.knn.plugin.transport.KNNStatsResponse;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.threadpool.Scheduler;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Service handling native memory circuit breaking logic. The circuit breaker gets tripped based on memory demand
+ * tracked by the {@link NativeMemoryCacheManager}. When {@link NativeMemoryCacheManager}'s cache fills up, if the
+ * circuit breaking logic is enabled, it will trip the circuit. Elsewhere in the code, the circuit breaker's value can
+ * be queried to prevent actions that should not happen during high memory pressure.
+ */
+public class NativeMemoryCircuitBreakerService extends AbstractLifecycleComponent {
+    private final ThreadPool threadPool;
+    private final KNNSettings knnSettings;
+    private final ClusterService clusterService;
+    private final Client client;
+    // Cancellable task to track circuitBreakerRunnable. In order to schedule, doStart must be called. doStart will
+    // only start the future if the previous value is null. Therefore, to close this class, do NOT set the value of
+    // this variable to null. To stop this class, this variable should be set to null so that it may be restarted.
+    @VisibleForTesting
+    final AtomicReference<Scheduler.Cancellable> circuitBreakerFuture;
+
+    public static final int CB_TIME_INTERVAL = 2 * 60; // seconds
+    private static final Logger logger = LogManager.getLogger(NativeMemoryCircuitBreakerService.class);
+
+    /**
+     * Constructor for creation of circuit breaker service for KNN
+     *
+     * @param knnSettings Settings class for k-NN
+     * @param threadPool thread pool for circuit breaker monitor to run job
+     * @param clusterService cluster service used to retrieve information about the cluster
+     * @param client client used to make calls to the cluster
+     */
+    public NativeMemoryCircuitBreakerService(KNNSettings knnSettings, ThreadPool threadPool, ClusterService clusterService, Client client) {
+        this.knnSettings = knnSettings;
+        this.threadPool = threadPool;
+        this.clusterService = clusterService;
+        this.client = client;
+        this.circuitBreakerFuture = new AtomicReference<>(null);
```

Comment:
Setting this explictly to null as it is a reference. For doStart, it will only execute if this value is null. doStop will reset the value to null, while doClose will not. So, this will allow the doStop/doStart logic to function properly.

Reply:
can we move this circuitBreakerFuture init in the doStart function? will that solves the problem?